### PR TITLE
Update MethodHandleResolver

### PR DESCRIPTION
**Pull request by babsingh**
1. In OJDK, [MHN.linkMethodHandleConstant](https://github.com/ibmruntimes/openj9-openjdk-jdk/blob/09e79a7ee9dce92e1741867f43e9b20c58e16f27/src/java.base/share/classes/java/lang/invoke/MethodHandleNatives.java#L602-L619) resolves a `MethodHandle`
constant pool entry. It is being incorporated in OpenJ9 to be used with
OJDK MHs. It has certain fast-paths which improve the performance.

2. Add inaccessible type checks in OJDK MHs' version of
`sendResolveMethodHandle` and `resolveInvokeDynamic`. This is equivalent to
`accessCheckArgRetTypes`. `inaccessibleTypeCheck` is a static version of
`accessCheckArgRetTypes`.

3. Remove the `try` block from OJDK MH's version of `resolveInvokeDynamic`.
It is not needed for Java 16. It incorrectly converts certain throwables
to different ones. Also, `MethodHandleNatives.linkCallSite` throws the
correct throwable. So, the throwable does not need to be caught and
converted into a different one. For JDK8, this method may need to be
re-visited due to spec differences. Currently, the spec differences
cannot be identified since OpenJ9 JDK8 cannot be built with OJDK MHs
enabled.

Fixes: #11934

Related: https://github.com/eclipse/openj9/issues/10913

Co-authored-by: Jack Lu <Jack.S.Lu@ibm.com>
Signed-off-by: Babneet Singh <sbabneet@ca.ibm.com>

### CODE REVIEW

**Comment by babsingh** [comment]
Recreated https://github.com/eclipse/openj9/pull/11371 since it cannot be reopened as per (https://github.com/eclipse/openj9/pull/11371#issuecomment-770015426). All the correspondence in https://github.com/eclipse/openj9/pull/11371 is important and related to this PR.

**Comment by babsingh** [comment]
fyi @DanHeidinga

**Comment by DanHeidinga** [comment]
Reminder - the key question from #11371 was whether the new OJDK-based MHs need the equivalent of `accessCheckArgRetTypes` access checks before this code is merged

**Comment by babsingh** [comment]
@DanHeidinga This PR is ready for review.

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandleResolver.java
Comment (DanHeidinga) [inline jcl/src/java.base/share/classes/java/lang/invoke/MethodHandleResolver.java:264]: Has this todo been addressed or recorded elsewhere?

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandleResolver.java
Comment (DanHeidinga) [inline jcl/src/java.base/share/classes/java/lang/invoke/MethodHandleResolver.java:356]: This seems like an expensive way to iterate the MT parameterArray() as it requires multiple allocations:
* one in `type.parameterArray()`
* one in `new Class[numElements]`
And an arraycopy.

Is this only to be able to iterate thru using the `for each` syntax?

This could be done by using `type.parameterArray()` and iterate thru it followed by checking the return type.  Or even better, use the package private `MethodType#Class<?>[] ptypes()` method to avoid the allocation entirely.

```
Class<?> clsArray = type.ptypes();  // or type.parameterArray()
for (Class<?> cls : clsArray) {
	if (!cls.isPrimitive()) {
		if (!lookup.isClassAccessible(cls)) {
			/*[MSG "K0680", "Class '{0}' no access to: class '{1}'"]*/
			throw new IllegalAccessError(Msg.getString("K0680", lookup.lookupClass(), cls));
		}
	}
}
Class<?> cls = type.returnType();
if (!cls.isPrimitive()) {
if (!lookup.isClassAccessible(cls)) {
	/*[MSG "K0680", "Class '{0}' no access to: class '{1}'"]*/
	throw new IllegalAccessError(Msg.getString("K0680", lookup.lookupClass(), cls));
	}
}
```
Better to avoid the allocation even if duplicating the code for the check.

Comment (babsingh) [reply jcl/src/java.base/share/classes/java/lang/invoke/MethodHandleResolver.java]: Updated.

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandleResolver.java
Comment (babsingh) [reply jcl/src/java.base/share/classes/java/lang/invoke/MethodHandleResolver.java:264]: It is recorded in as item number (2) in https://github.com/eclipse/openj9/issues/10913. I have also re-added the comment in the PR.

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandleResolver.java
Comment (DanHeidinga) [inline jcl/src/java.base/share/classes/java/lang/invoke/MethodHandleResolver.java:259]: nitpick - extra `;`
```suggestion
		result[0] = MethodHandleNatives.linkCallSite(classObject, 0, bsm, name, type, (Object)staticArgs, appendixResult);
```

**Review by DanHeidinga — APPROVED** [review]
Lgtm - let me know if you want to fix the nitpick or not and then I'll launch builds

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandleResolver.java
Comment (babsingh) [reply jcl/src/java.base/share/classes/java/lang/invoke/MethodHandleResolver.java]: fixed.

**Comment by DanHeidinga** [comment]
Jenkins test sanity zlinux jdk8,jdk11,jdk16

**Comment by babsingh** [comment]
**JDK8 and JDK11** builds failed since [Lookup.isClassAccessible](https://github.com/eclipse/openj9/blob/9530f964a5f4a1a95a0e5147a462af8e9b987a29/jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java#L1450-L1463) (within OpenJ9 `MethodHandles`) is only available in JDK14+. 

**OJDK11** `MethodHandles` has `Lookup.isClassAccessible`: https://github.com/ibmruntimes/openj9-openjdk-jdk11/blob/5778b50c73deb9b07dfd769566aa6631074c64d8/src/java.base/share/classes/java/lang/invoke/MethodHandles.java#L2058

**OJDK8** `MethodHandles` does not have `Lookup.isClassAccessible`: https://github.com/ibmruntimes/openj9-openjdk-jdk8/blob/openj9/jdk/src/share/classes/java/lang/invoke/MethodHandles.java

I can expose `Lookup.isClassAccessible` to JDK8 and JDK11 in OpenJ9. This will temporarily resolve the issue. But, OJDK8 `MethodHandles` does not have `Lookup.isClassAccessible`. So, `Lookup.isClassAccessible` will need to be replaced with an alternative before enabling OJDK MHs in OpenJ9 JDK8. Also, I noticed that OJDK and OJ9 `Lookup.isClassAccessible` differ; the OJ9 version does a security check, which is not performed in the OJDK version.

I will try to find a better solution.

**Comment by babsingh** [comment]
> OJDK8 MethodHandles does not have `Lookup.isClassAccessible`

[FIX] The definition of `Lookup.isClassAccessible` (package protected method) will be backported from OJDK11 to OJDK8 extensions repo.

> OJDK and OJ9 `Lookup.isClassAccessible` differ; the OJ9 version does a security check, which is not performed in the OJDK version

[FIX] OJ9 versions of `sendResolveMethodHandle` and `resolveInvokeDynamic` will continue to use `accessCheckArgRetTypes` instead of `inaccessibleTypeCheck`.

This PR has been updated to reflect the above changes. @DanHeidinga Can you please review this PR?

**Comment by DanHeidinga** [comment]
Jenkins test sanity zlinux jdk8,jdk11,jdk16

**Comment by DanHeidinga** [comment]
Assuming the builds pass, this is good to be merged

**Comment by babsingh** [comment]
@DanHeidinga the builds have passed. can this be merged?

**Pull request merged**
